src/mmtk.rs

```diff
@@ -467,4 +469,46 @@ impl<VM: VMBinding> MMTK<VM> {
     pub fn get_options(&self) -> &Options {
         &self.options
     }
+
+    /// Enumerate objects in all spaces in this MMTK instance.
+    ///
+    /// The call-back function `f` is called for every object that has the valid object bit (VO
+    /// bit), i.e. objects that are allocated in the heap of this MMTK instance, but has not been
+    /// reclaimed, yet.
+    ///
+    /// # Interaction with allocation and GC
+    ///
+    /// This function does not mutate the heap.  It is safe if multiple threads execute this
+    /// function concurrently during mutator time.
+    ///
+    /// This function will visit all objects that have been allocated at the time when this function
+    /// is called.  But if new objects are allocated while this function is being executed, this
+    /// function may or may not visit objects allocated after this function started.
+    ///
+    /// Also note that when this function visits an object, it only guarantees that its VO bit must
+    /// have been set.  It is not guaranteed if the object has been "fully initialized" in the sense
+    /// of the programming language the VM is implementing.  For example, the object header and the
+    /// type information may not have been written.
+    ///
+    /// It has *undefined behavior* if GC happens while this function is being executed.  The VM
+    /// binding must ensure GC does not start while executing this function.  One way to prevent GC
+    /// from starting is not letting the current thread yield for GC.
+    ///
+    /// Some programming languages may provide an API that allows the user to allocate objects and
+    /// trigger GC while enumerating objects.  The VM binding may use the callback of this function
+    /// to save all visited objects in an array and let the user visit the array after this function
+    /// returned.
```

Comment:
I don't see how it would be helpful to save the visited objects and visit them later. If the GC is happening, the visited objects may be reclaimed during this call, and the saved object references may become invalid.

**As this function uses VO bit (which is changed during GC and allocation), and the function does not block GC and does not block allocation, it should be undefined behavior if GC or allocation happens at the same time.** We cannot guarantee anything if GC/allocation is happenning. We may call the closure for an object that seems valid, but when the closure visits the object, the object may be reclaimed. We may visit objects that are copied (but their VO bits are not yet cleared). We may miss objects that are just allocated.

Reply:
> I don't see how it would be helpful to save the visited objects and visit them later. If the GC is happening, the visited objects may be reclaimed during this call, and the saved object references may become invalid.

The VM binding needs to treat the saved list of object references as roots.  In the way, if GC happens, those objects will be guaranteed to live through the next GC.  The list may contain the last reference to some objects because the object may be unlinked from their original parents after the last GC.  That's expected.  Such heap traversal APIs are allowed to pick up objects that have become unreachable, finalized, etc., and the VM binding needs to be aware of that.  But if such unreachable objects points to other objects, their outgoing edges will remain valid because (1) if GC has not happened, their edges remain valid, and (2) if GC happens, those objects are rooted and will be traced and forwarded.

If the VM can update the roots (in the same way updating JNI handles or other native roots), the saved list can be updated if a copying GC happens.  Alternatively, if the VM only intends to support plans that support pinning, it can treat the list as pinning roots, and those objects will not be moved while iterating through the list.

I have tested it on Ruby by manually triggering a GC before saving the list (as a root) and iterating through the list.  It works.

Reply:
What you said does not contradict with what I said. I suggested that 'it should be undefined behavior if GC or allocation happens at the same time (during the enumerating function call)'. After the enumerating function, the binding can do whatever they need to make an 'illusion' that the user is enumerating objects during a GC. But this function itself is not allowed during GC.
```suggestion
    /// trigger GC while enumerating objects.  The VM binding may use the callback of this function
    /// to save all visited objects in an array outside GCs and let the user visit the array after this function
    /// returned. The binding needs to carefully keep the object references in the array alive and updated properly if
    /// a GC happens.
```

Reply:
I agree that it should have undefined behavior if GC happens.

I thought allocation should be benign because allocation strictly adds VO bits but does not remove VO bits.  But after thinking about it, I find that it is easier to just specify that it has undefined behavior if allocation happens concurrently, too.  Reasons include

-   We currently accesses the VO bits at different granularities, including bit, byte, word, etc.  The interaction of such memory accesses in a multi-threaded environment is not well-defined in the C++11 or Rust memory model.  So it is hard to reason what the reader will see (or whether it has undefined behavior) if the read and the write happens concurrently.
-   In a multi-threaded program, different threads may not agree upon a global total order in which different VO bits are set.  (Thread 1 may see bit X is set before bit Y, but Thread2 may see bit Y is set before bit X.)  So "all bits that have been set when `enumerate_objects` is called" is not precisely defined unless we make all threads stop, or at least synchornize with the current thread.

It is OK for Ruby because we currently only supports a single Ractor (in which only one thread is active at any time).  For OpenJDK, it will require a `VMOperation` to stop the world before enumerating objects.

Reply:
I updated the comment.  I made it clear that it is a data race if either allocation or GC happens concurrently with `enumerate_objects`.  I also added some comments on keeping the saved object references in roots if the VM binding implements an API that allows allocation while enumerating objects.